docs/source/howto/data.rst

```diff
@@ -248,11 +247,247 @@ The node class has various methods to set these attributes, such as :py:`~aiida.
+    qb.append(CalcJobNode)    # Setting first vertex of path
+
+If you are interested in instances of different classes, you can also pass a tuple, list or set of classes.
+However, they have to be of the same ORM-type (e.g. all have to be subclasses of ``Node``):
+
+.. code-block:: python
+
+    qb = QueryBuilder()       # Instantiating instance. One instance -> one query
+    qb.append([CalcJobNode, WorkChainNode]) # Setting first vertice of path, either WorkChainNode or Job.
+
+.. note::
+
+    Remember that :ref:`processes<topics:processes:concepts:types>` have both a run-time ``Process`` that steers them and a ``Node`` that stores their metadata in the database.
+    The QueryBuilder allows you to pass either the ``Node`` class (e.g. ``CalcJobNode``) or the ``Process`` class (e.g. ``CalcJob``), which will automatically apply the correct filters for the type of calculation.
+
+.. _how-to:data:find:results:
+
+Retrieving results
+------------------
+
+Once you have *appended* the entity you want to query for to the ``QueryBuilder``, the next question is how to get the results.
+There are several ways to obtain data from a query:
+
+.. code-block:: python
+
+    qb = QueryBuilder()                 # Instantiating instance
+    qb.append(CalcJobNode)              # Setting first vertice of path
+
+    first_row = qb.first()              # Returns a list (!)
+                                        # of the results of the first row
+
+    all_results_d = qb.dict()           # Returns all results as
+                                        # a list of dictionaries
+
+    all_results_l = qb.all()            # Returns a list of lists
+
+In case you are working with a large dataset, you can also return your query as a generator:
+
+.. code-block:: python
+
+    all_res_d_gen = qb.iterdict()       # Return a generator of dictionaries
+                                        # of all results
+    all_res_l_gen = qb.iterall()        # Returns a generator of lists
+
+This will retrieve the data in batches, and you can start working with the data before the query has completely finished.
+
+.. _how-to:data:find:filters:
+
+Filters
+-------
+
+Usually you do not want to query for *all* entities of a certain class, but rather *filter* the results based on certain properties.
+Suppose you do not want to all ``CalcJobNode`` data, but only the ones that are ``finished``:
+
+.. code-block:: python
+
+    qb = QueryBuilder()                 # Initialize a QueryBuilder instance
+    qb.append(
+        CalcJobNode,                    # Append a CalcJobNode
+        filters={                       # Specify the filters:
+            'attributes.process_state': 'finished',  # the process is finished
+        },
+    )
+
+You can apply multiple filters to one entity in a query.
+Say you are interested in all calculation jobs in your database that are ``finished`` **and** have the ``process_label`` ``ArithmeticAddCalculation``:
```

Comment:
That's a good point, but aren't all processes with a numerical `exit_status` finished? So in case you specify the `'attributes.exit_status': 0` filter, isn't the `'attributes.process_state': 'finished'` filter redundant?

Maybe we can use `'attributes.resources.num_machines': 1` as the second filter, or is this too complicated?

Reply:
> That's a good point, but aren't all processes with a numerical `exit_status` finished?

Yes, that is true. But I think the "hidden inefficiency" does not outweigh the advantage of the clarity of the example. I indeed think the resources alternative may be a bit too complicated.